pallets/parachain-system/src/validate_block/implementation.rs

```diff
@@ -77,39 +77,41 @@ pub fn validate_block<
 	PSC: crate::Config,
 	CI: crate::CheckInherents<B>,
 >(
-	params: ValidationParams,
+	MemoryOptimizedValidationParams {
+		block_data,
+		parent_head,
+		relay_parent_number,
+		relay_parent_storage_root,
+	}: MemoryOptimizedValidationParams,
 ) -> ValidationResult
 where
 	B::Extrinsic: ExtrinsicCall,
 	<B::Extrinsic as Extrinsic>::Call: IsSubType<crate::Call<PSC>>,
 {
-	let block_data = ParachainBlockData::<B>::decode(&mut &params.block_data.0[..])
+	let block_data = codec::decode_from_bytes::<ParachainBlockData<B>>(block_data)
 		.expect("Invalid parachain block data");
 
-	let parent_head =
-		B::Header::decode(&mut &params.parent_head.0[..]).expect("Invalid parent head");
-
-	// We decoded the block data and don't need it anymore. So, we release it to free the memory.
-	mem::drop(params.block_data);
+	let parent_header =
+		codec::decode_from_bytes::<B::Header>(parent_head.clone()).expect("Invalid parent head");
 
 	let (header, extrinsics, storage_proof) = block_data.deconstruct();
 
 	let head_data = HeadData(header.encode());
```

Comment:
I think this allocation could be delayed? (e.g. by doing `let head_data = HeadData(block.header().encode())` since `block` is alive until the very end).